core/src/main/scala/kafka/zookeeper/ZookeeperClient.scala

```diff
@@ -279,33 +258,130 @@ trait ZNodeChildChangeHandler {
 }
 
 sealed trait AsyncRequest {
-  val path: String
-  val ctx: Any
+  type Response <: AsyncResponse
+  def path: String
+  def ctx: Option[Any]
+
+  protected[zookeeper] def send(zooKeeper: ZooKeeper)(processResponse: Response => Unit)
+}
+
+sealed trait WatchableAsyncRequest extends AsyncRequest {
+
+  final protected[zookeeper] def send(zooKeeper: ZooKeeper)(processResponse: Response => Unit) =
```

Comment:
It seems that this variant of send in `WatchableAsyncRequest` should never get called.

Since this PR is already going down the rabbit hole of solving issues through types, you could basically split AsyncRequest into watchable/non-watchable sealed traits, remove the `send` api from the AsyncRequest layer, and push exactly the right signatures down into the watchable/non-watchable sealed traits.

Reply:
If it's not meant to be called, perhaps we could just throw an IllegalStateException?